keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,509 @@
+
+Nevertheless, risk mitigation has been discussed in details (see archived options below) and we agree to start with the following node metrics of cpu pool sizes in Alpha stage to assess the actual impact in real deployment before revisiting if we need risk mitigation.
+
+https://github.com/kubernetes/kubernetes/pull/127506
+- report shared pool size, in millicores (e.g. 13500m)
+- report exclusively allocated cores, counting full cores (e.g. 16)
+
+
+#### Archived Risk Mitigation (Option 1)
+
+This option is to add `numMinSharedCPUs` in `strict-cpu-reservation` option as the minimum number of CPU cores not available for exclusive allocation and expose it to Kube-scheduler for enforcement.
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
+
+![MinSharedCPUs](./strict-cpu-allocation.png)
+
+When `strict-cpu-reservation` is disabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+defaultCPUSet = Reserved (6) + 58 (available for exclusive allocation)
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+MinSharedCPUs: 4
+defaultCPUSet = MinSharedCPUs (4) + 54 (available for exclusive allocation)
+```
+
+Prototype PR for the option is created:
+https://github.com/kubernetes/kubernetes/pull/123979/commits
+
+Add `numMinSharedCPUs` as part of `strict-cpu-reservation` option in Kubelet configuration:
+
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+featureGates:
+  ...
+  CPUManagerPolicyOptions: true
+  CPUManagerPolicyAlphaOptions: true
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  strict-cpu-reservation: { "enable": "true", "numMinSharedCPUs": 4 }
+reservedSystemCPUs: "0,32,1,33,16,48"
+...
+```
+
+In Node API, we add `exclusive-cpu` in Node Allocatable for Kube-scheduler to consume.
+
+```
+  "status": {
+    "capacity": {
+      "cpu": "64",
+      "exclusive-cpu": "64",
+      "ephemeral-storage": "832821572Ki",
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "196146004Ki",
+      "pods": "110"
+    },
+    "allocatable": {
+      "cpu": "58",
+      "exclusive-cpu": "54",
+      "ephemeral-storage": "767528359485",
+      "hugepages-1Gi": "0",
+      "hugepages-2Mi": "0",
+      "memory": "186067796Ki",
+      "pods": "110"
+    },
+  ...
+```
+
+In kube-scheduler, `ExlusiveMilliCPU` is added in scheduler's `Resource` structure and `NodeResourcesFit` plugin is extended to filter out nodes that can not meet pod's exclusive CPU request.
```

Comment:
I have a feeling, that this part of discussion is part of another problem, not the one that is covered in goals for this KEP or in user stories. "Can the workload get exclusive CPU on particular node" is separate from "Let's not use reservedCPUs for workloads, as they were meant for host OS system services".

Reply:
agreed, this option is archived for now.